### PR TITLE
Added Normalization to Kets and DensityMatrix (Added Checks for Hermiticity and Trace >1)

### PR DESCRIPTION
**Pull request by aabhasgulati**
Added Normalization to Kets and DensityMatrix (Added Checks for Hermiticity and Trace >1)
Also exports the normalization function for Kets and Bras!
You are welcome!

### CODE REVIEW

File: src/QuantumObjects.jl [hunk 1]
```diff
@@ -1,18 +1,26 @@
...
+    function Ket(coefficients)
+        new(coefficients ./ norm(coefficients))
+    end  
 end
 
 struct Bra <: QuantumObject
     k::Ket
 end
 
+norm(k1:: Ket) = norm(k1.coefficients)
```
Comment (DhruvaSambrani) [inline src/QuantumObjects.jl:21]: Dont find norm again, just set to 1.

Remove space after `::`

File: src/QuantumObjects.jl [hunk 2]
```diff
@@ -23,6 +31,17 @@ Base.:*(b::Bra, k::Ket) = inner(Ket(b), k)
 
 struct DensityMatrix <: QuantumObject
     matrix::Hermitian{ComplexF64, Matrix{ComplexF64}}
+    function DensityMatrix(rho::Matrix{ComplexF64}, to_normalize=true:: Bool)
+        if rho != adjoint(rho)
```
Comment (DhruvaSambrani) [inline src/QuantumObjects.jl:35]: Use ishermitian instead

File: src/QuantumObjects.jl [hunk 3]
```diff
@@ -23,6 +31,17 @@ Base.:*(b::Bra, k::Ket) = inner(Ket(b), k)
...
+            error("Density Matrix has to be Hermitian!")
+        end
+        if !to_normalize & (real(tr(rho)) > 1.0)
+            error("Density Matrix should have Trace smaller than or equal to  1")
+        elseif to_normalize 
+            rho = rho ./ tr(rho)
+        end
+            new(Hermitian(rho))
+    end
 end
```
Comment (DhruvaSambrani) [inline src/QuantumObjects.jl:46]: Define norm for DM

**Review by DhruvaSambrani — CHANGES_REQUESTED** [review]

File: src/QuantumObjects.jl [hunk 4]
```diff
@@ -1,18 +1,26 @@
 module QuantumObjects
 import Base: *, ∘, kron, adjoint
 import LinearAlgebra: Hermitian, tr
-export QuantumObject, Ket, Bra, DensityMatrix, inner
+export QuantumObject, Ket, Bra, DensityMatrix, inner,norm
 
 abstract type QuantumObject end
 
+norm(coefficients::Array{ComplexF64,1})  = sqrt(sum((abs.(coefficients)).^2))
```
Comment (DhruvaSambrani) [inline src/QuantumObjects.jl:8]: Do not redefine function from LinAlg

**Review by DhruvaSambrani — CHANGES_REQUESTED** [review]

File: src/QuantumObjects.jl (hunk 3 above)
Comment (DhruvaSambrani) [reply src/QuantumObjects.jl:46]: Norm is tr(rho^2)

**Comment by aabhasgulati** [comment]
I have made all the required changes.

File: src/QuantumObjects.jl (hunk 4 above)
Comment (DhruvaSambrani) [reply src/QuantumObjects.jl:8]: This is not resolved. 
Import norm from LinearAlgebra, add a method definition for the different objects.

**Pull request merged**